janus_server/src/bin/aggregation_job_driver.rs

```diff
@@ -0,0 +1,1422 @@
+                    // tightly loop running transactions that will fail without any delay.
+                    job_discovery_delay = self.step_job_discovery_delay(job_discovery_delay);
+                    continue;
+                }
+            };
+            if acquired_jobs.is_empty() {
+                debug!("No aggregation jobs available");
+                job_discovery_delay = self.step_job_discovery_delay(job_discovery_delay);
+                continue;
+            }
+            info!(
+                acquired_job_count = acquired_jobs.len(),
+                "Acquired aggregation jobs"
+            );
+
+            // Start up tasks for each acquired aggregation job.
+            job_discovery_delay = Duration::ZERO;
+            for (task_id, vdaf, aggregation_job_id, lease_expiry) in acquired_jobs {
+                info!(
+                    ?task_id,
+                    ?aggregation_job_id,
+                    ?lease_expiry,
+                    "Stepping aggregation job"
+                );
+                task::spawn({
+                    // We acquire a semaphore in the job-discovery task rather than inside the new
+                    // job-stepper task to ensure that acquiring a permit does not race with
+                    // checking how many permits we have available in the next iteration of this
+                    // loop, to maintain the invariant that this task is the only place we acquire
+                    // permits.
+                    //
+                    // Unwrap safety: we have seen that at least `acquired_jobs.len()` permits are
+                    // available, and this task is the only task that acquires permits.
+                    let permit = Arc::clone(&sem).try_acquire_owned().unwrap();
+                    let this = Arc::clone(&self);
+                    async move {
+                        match time::timeout(
+                            this.effective_lease_duration(lease_expiry),
+                            this.step_aggregation_job(vdaf, task_id, aggregation_job_id),
+                        )
+                        .await
+                        {
+                            Ok(Ok(_)) => {
+                                debug!(?task_id, ?aggregation_job_id, "Aggregation job stepped")
+                            }
+
+                            Ok(Err(err)) => {
+                                error!(
+                                    ?task_id,
+                                    ?aggregation_job_id,
+                                    ?err,
+                                    "Couldn't step aggregation job"
+                                )
+                            }
+
+                            Err(err) => error!(
+                                ?task_id,
+                                ?aggregation_job_id,
+                                ?err,
+                                "Couldn't step aggregation job"
```

Comment:
Sure, we can -- I was sorta viewing these two error cases as the "same" error handler, split out for purely technical reasons -- but since we're splitting I suppose there's nothing wrong with varying the error message, too.